feat(web-analytics): query builder for pre-aggregated trends

joshsny opened the pull request:
## Problem

We want to build trends queries on pre-aggregated data

## Changes

Adds a query builder for trends using pre-aggregated tables

## How did you test this code?

Unit tests

---

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

joshsny committed:
feat: query builder for trends on pre-aggregated tables

joshsny committed:
fix mypy errors in tests

robbie-c commented on posthog/hogql_queries/web_analytics/test/test_trends_pre_aggregated_query_builder.py:
I was going to say, we need some tests here the queries return the correct data

joshsny committed:
generate schema rather than editing directly

joshsny committed:
Merge branch 'master' into web-analytics-trends-query-builder

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

joshsny committed:
Merge branch 'master' into web-analytics-trends-query-builder

robbie-c commented on posthog/schema.py:
```diff
@@ -2505,6 +2506,35 @@ class WebStatsBreakdown(StrEnum):
+    model_config = ConfigDict(
+        extra="forbid",
+    )
+    Bounces: Optional[float] = None
```
Let's call this BounceRate, we want to show the avg(is_bounce), rather than sum(is_bounce)

robbie-c commented on posthog/hogql_queries/web_analytics/trends_pre_aggregated_query_builder.py:
```diff
@@ -0,0 +1,145 @@
+            expr=bucket_expr,
+        )
+
+    def _get_metric_expr(self, metric: WebTrendsMetric) -> ast.Alias:
```
There's a limitation in the existing Trends query runner that stops us from showing trends of Bounce Rate over time.

Trends don't support averages across Sessions, only across Events, so things like average session duration and bounce rate don't work (sessions with more events would contribute to the average multiple times). If we're writing our own trends alternative, we can implement those metrics, which would be awesome!

lricoy replied:
Yep, same for session duration. This has come up before, and it will be great to add them!

lricoy committed:
Merge branch 'master' into web-analytics-trends-query-builder

lricoy reviewed (approved):
I will go ahead and merge this so I can open another one with the changes mentioned and the new response format.

Thanks for doing that!

lricoy committed:
chore: add WebTrendsQueryRunner implementation for pre-aggregated trends

- Implement WebTrendsQueryRunner with full pre-aggregated table support
- Add comprehensive test suite covering query generation and metrics
- Update query dispatch system to handle WebTrendsQuery routing
- Regenerate Python schema to include new WebTrendsQuery types
- Add WebTrendsQuery to RunnableQueryNode union for proper type checking

lricoy committed:
Fix import sorting

lricoy committed:
fix merge conflicts

- Integrated WebTrendsQuery import with master branch changes
- Cleaned up duplicate imports from merge conflict
- Maintained proper import organization and structure

lricoy committed:
fix import sort

lricoy committed:
Merge branch 'master' into web-analytics-trends-query-builder

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

lricoy committed:
enforce new mypy rules

lricoy committed:
Update mypy baseline after fixing typing issues

lricoy committed:
fix: missing query metada

lricoy committed:
chore: update schema.py after query changes

lricoy committed:
chore: trigger CI

lricoy committed:
Merge branch 'master' into web-analytics-trends-query-builder

lricoy committed:
Fix schema generation inconsistencies

Update both frontend and Python schemas to match latest generated output:

Frontend (schema.json):
- Add query_metadata field to cached query response interfaces
- Update generated interface hash

Python (schema.py):
- Add query_metadata field to CachedWebTrendsQueryResponse
- Update generated interface hash

This ensures CI schema build checks pass by keeping schema files
in sync with current TypeScript definitions.
